zzk/service/service_test.go

```diff
@@ -97,3 +97,88 @@ func (t *ZZKTest) TestWaitService(c *C) {
 		c.Fatalf("Timed out waiting for listener")
 	}
 }
+
+func (t *ZZKTest) TestWaitInstance(c *C) {
+
+	node := &ServiceNode{
+		ID:           "serviceid",
+		Name:         "serviceA",
+		DesiredState: int(service.SVCRun),
+		Instances:    1,
+	}
+
+	conn, err := zzk.GetLocalConnection("/")
+	c.Assert(err, IsNil)
+
+	// add 1 service
+	err = conn.CreateDir("/pools/poolid/services/serviceid")
+	c.Assert(err, IsNil)
+
+	// add an node for an instance of that service
+	err = conn.Create("/pools/poolid/services/serviceid/hostid-serviceid-0", node)
+	c.Assert(err, IsNil)
+
+	// add 1 host
+	err = conn.CreateDir("/pools/poolid/hosts/hostid/instances")
+	c.Assert(err, IsNil)
+
+	// add an node for an instance of that service
+	err = conn.Create("/pools/poolid/hosts/hostid/instances/hostid-serviceid-0", node)
+	c.Assert(err, IsNil)
+
+	instanceID := 0
+
+	// Start our wait and check/return when the state is SVCRun and started is after terminated
+	shutdown := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
+		checkState := func(s *State, ok bool) bool {
+			if !ok {
+				return false
+			}
+			return s.DesiredState == service.SVCRun && s.Started.After(s.Terminated)
+		}
+
+		err := WaitInstance(shutdown, conn, "poolid", "serviceid", instanceID, checkState)
+		c.Assert(err, IsNil)
+		close(done)
+	}()
+
+	timer := time.NewTimer(time.Second)
+	defer timer.Stop()
+
+	select {
+	case <-done:
+		c.Fatalf("Listener exited unexpectedly")
+	case <-timer.C:
+	}
+
+	// create the state
+	req := StateRequest{
+		PoolID:     "poolid",
+		HostID:     "hostid",
+		ServiceID:  "serviceid",
+		InstanceID: instanceID,
+	}
+
+	timer.Reset(time.Second)
```

Comment:
WaitInstance isn't going to return until UpdateState finishes, then it will `close(done)`.  We want to see this happen, but within some short time period, so we fail if there is a bug in WaitInstance where it isn't returning in a timely manner.